src/openmc_heat_driver.cpp

```diff
@@ -100,36 +125,143 @@ void OpenmcHeatDriver::init_mappings()
+  // can take a point on a 45 degree ray from the pin center. TODO: add a check to make
+  // sure that the T/H model is finer than the OpenMC model.
+  for (int i = 0; i < heat_driver_->n_pins_; ++i) {
+    double x_center = heat_driver_->pin_centers_(i, 0);
+    double y_center = heat_driver_->pin_centers_(i, 1);
+
+    for (int j = 0; j < heat_driver_->n_axial_; ++j) {
+      double zavg = 0.5 * (z(j) + z(j + 1));
+      double l = heat_driver_->pin_pitch() / std::sqrt(2.0);
+      double d = (l - heat_driver_->clad_outer_radius_) / 2.0;
+      double x = x_center + (heat_driver_->clad_outer_radius_ + d) * std::sqrt(2.0) / 2.0;
+      double y = y_center + (heat_driver_->clad_outer_radius_ + d) * std::sqrt(2.0) / 2.0;
+
+      // Determine cell instance corresponding to given fluid location
+      Position r{x, y, zavg};
+      CellInstance c{r};
+      if (tracked.find(c) == tracked.end()) {
+        openmc_driver_->cells_.push_back(c);
+        tracked[c] = openmc_driver_->cells_.size() - 1;
+
+        Ensures(!c.is_fissionable());
+      }
+
+      // Map OpenMC material to fluid element and vice versa
+      int32_t array_index = tracked[c];
+      cell_inst_to_elem_[array_index].push_back(elem_index);
+      elem_to_cell_inst_[elem_index].push_back(array_index);
+
+      ++elem_index;
+    }
+  }
+
+  if (openmc_driver_->active()) {
+    n_fluid_cells_ = openmc_driver_->cells_.size() - n_solid_cells_;
   }
 }
 
 void OpenmcHeatDriver::init_tallies()
 {
   if (openmc_driver_->active()) {
-    // Build vector of material indices
+    // Build vector of material indices to construct tallies; tallies are only
+    // used in the solid regions
     std::vector<int32_t> mats;
-    for (const auto& c : openmc_driver_->cells_) {
-      mats.push_back(c.material_index_);
+    for (int c = 0; c < n_solid_cells_; ++c) {
+      const auto& cell = openmc_driver_->cells_[c];
+      mats.push_back(cell.material_index_);
     }
+
     openmc_driver_->create_tallies(mats);
   }
 }
 
+void OpenmcHeatDriver::init_densities()
+{
+  if (this->has_global_coupling_data()) {
+    std::size_t n_solid =
+      heat_driver_->n_pins_ * heat_driver_->n_axial_ * heat_driver_->n_rings();
+    std::size_t n_fluid = heat_driver_->n_pins_ * heat_driver_->n_axial_;
+    densities_.resize({n_solid + n_fluid});
+    densities_prev_.resize({n_solid + n_fluid});
+
+    if (density_ic_ == Initial::neutronics) {
+      // Loop over all of the fluid regions in the heat transfer model and set
+      // the density IC based on the densities used in the OpenMC input file. Set
+      // the initial solid density to 0.0 because it is not used in the simulation
+      // at all anyways.
+      int ring_index = 0;
+      for (int i = 0; i < heat_driver_->n_pins_; ++i) {
+        for (int j = 0; j < heat_driver_->n_axial_; ++j) {
+          for (int k = 0; k < heat_driver_->n_rings(); ++k, ++ring_index) {
+            densities_(ring_index) = 0.0;
```

Comment:
This whole loop would be unnecessary if `densities_` was just initialized to zeros.
```C++
densities_.fill(0.0);
```